### PR TITLE
BetterReflection entry-point class

### PR DESCRIPTION
**Pull request by Ocramius**
This patch introduces an "entry point" that can be used to get started with `BetterReflection`.

The usage is as following:

```php
$class = (new \Roave\BetterReflection\Configuration)->classReflector()->reflect(ClassName::class)
```

This also allows us to have clean DI on each of the internal components, since the orchestration is organised by this class.

I will provide detailed info about the BC breaks and decisions to be taken on this below

### CODE REVIEW

File: demo/basic-reflection/example3.php [hunk 1]
```diff
@@ -4,12 +4,13 @@
...
 require_once __DIR__ . '/../../vendor/autoload.php';
 
+use Roave\BetterReflection\Configuration;
 use Roave\BetterReflection\Reflector\ClassReflector;
 use Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocator;
 use Roave\BetterReflection\SourceLocator\Type\SingleFileSourceLocator;
 
 $reflector = new ClassReflector(new AggregateSourceLocator([
-    new SingleFileSourceLocator(__DIR__ . '/assets/MyClass.php'),
+    new SingleFileSourceLocator(__DIR__ . '/assets/MyClass.php', (new Configuration())->astLocator()),
```
Comment (Ocramius) [inline demo/basic-reflection/example3.php]: As you can see, using a locator directly can be more tedious now. The various `SourceLocators` now require an `Ast\Locator` (and sometime a `Parser`) instance to be built

File: src/Configuration.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            ?? $this->compileNodeToValue = new CompileNodeToValue();
+    }
+
+    public function findReflectionsOnLine() : FindReflectionOnLine
+    {
+        return $this->findReflectionOnLine
+            ?? $this->findReflectionOnLine = new FindReflectionOnLine($this->sourceLocator(), $this->astLocator());
+    }
+
+    // @TODO modifier methods here. They should be `withFoo()`, `withBar()`, and be immutable.
```
Comment (Ocramius) [inline src/Configuration.php]: Don't know if we should make this configurable at all, at this point. After implementing the entire setup in the test suite, this degree of configuration was no longer needed.

File: src/Configuration.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+use Roave\BetterReflection\SourceLocator\Ast\Strategy\AstConversionStrategy;
+use Roave\BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection;
+use Roave\BetterReflection\SourceLocator\Type\AggregateSourceLocator;
+use Roave\BetterReflection\SourceLocator\Type\AutoloadSourceLocator;
+use Roave\BetterReflection\SourceLocator\Type\EvaledCodeSourceLocator;
+use Roave\BetterReflection\SourceLocator\Type\PhpInternalSourceLocator;
+use Roave\BetterReflection\SourceLocator\Type\SourceLocator;
+use Roave\BetterReflection\Util\FindReflectionOnLine;
+
+// @TODO should the naming be swapped around? `new BetterReflection()` sounds MUCH cooler!
```
Comment (Ocramius) [inline src/Configuration.php]: Currently, the logic is `new \Roave\BetterReflection\Configuration()`. In my opinion, `new \Roave\BetterReflection()` would be quite awesome

File: src/Configuration.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+final class Configuration implements BetterReflection
```
Comment (Ocramius) [inline src/Configuration.php]: After finishing it, I saw that the interface is not really needed. In my opinion, this is OK as-is, so the `implements` keyword should disappear.

File: src/Reflector/FunctionReflector.php [hunk 5]
```diff
@@ -15,9 +15,15 @@ class FunctionReflector implements Reflector
      */
     private $sourceLocator;
 
-    public function __construct(SourceLocator $sourceLocator)
+    /**
+     * @var ClassReflector
+     */
+    private $classReflector;
+
+    public function __construct(SourceLocator $sourceLocator, ClassReflector $classReflector)
```
Comment (Ocramius) [inline src/Reflector/FunctionReflector.php]: BC Break to be documented

File: src/SourceLocator/Ast/Locator.php [hunk 6]
```diff
@@ -28,11 +28,11 @@ class Locator
      */
     private $parser;
 
-    public function __construct()
+    public function __construct(Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Locator.php]: BC Break to be documented

File: src/SourceLocator/Reflection/SourceStubber.php [hunk 7]
```diff
@@ -28,9 +29,9 @@
      */
     private $prettyPrinter;
 
-    public function __construct()
+    public function __construct(Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Reflection/SourceStubber.php]: BC Break to be documented

File: src/SourceLocator/Type/AbstractSourceLocator.php [hunk 8]
```diff
@@ -31,9 +32,9 @@
      */
     abstract protected function createLocatedSource(Identifier $identifier) : ?LocatedSource;
 
-    public function __construct(?AstLocator $astLocator = null)
+    public function __construct(AstLocator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 9]
```diff
@@ -39,18 +40,20 @@
 
     /**
      * @param object $anonymousClassObject
+     * @param Parser $parser
      *
      * @throws \InvalidArgumentException
      * @throws \ReflectionException
      */
-    public function __construct($anonymousClassObject)
+    public function __construct($anonymousClassObject, Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/AutoloadSourceLocator.php [hunk 10]
```diff
@@ -17,6 +19,24 @@
...
     /**
+     * @var AstLocator
+     */
+    private $astLocator;
+
+    /**
+     * Note: the constructor has been made a 0-argument constructor because `\stream_wrapper_register`
+     *       is a piece of trash, and doesn't accept instances, just class names.
+     */
+    public function __construct(AstLocator $astLocator = null)
```
Comment (Ocramius) [inline src/SourceLocator/Type/AutoloadSourceLocator.php]: BC Break to be documented. Also note the comment above

File: src/SourceLocator/Type/ClosureSourceLocator.php [hunk 11]
```diff
@@ -38,10 +38,10 @@
      */
     private $parser;
 
-    public function __construct(Closure $closure)
+    public function __construct(Closure $closure, Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Type/ClosureSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/ComposerSourceLocator.php [hunk 12]
```diff
@@ -22,9 +23,9 @@ class ComposerSourceLocator extends AbstractSourceLocator
      */
     private $classLoader;
 
-    public function __construct(ClassLoader $classLoader)
+    public function __construct(ClassLoader $classLoader, Locator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/ComposerSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/DirectoriesSourceLocator.php [hunk 13]
```diff
@@ -23,15 +24,25 @@ class DirectoriesSourceLocator implements SourceLocator
...
+    private $astLocator;
+
+    /**
      * @param string[] $directories directories to scan
      *
      * @throws InvalidDirectory
      * @throws InvalidFileInfo
      */
-    public function __construct(array $directories)
+    public function __construct(array $directories, Locator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/DirectoriesSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/EvaledCodeSourceLocator.php [hunk 14]
```diff
@@ -17,10 +20,11 @@
      */
     private $stubber;
 
-    public function __construct()
+    public function __construct(Locator $astLocator, Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Type/EvaledCodeSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/FileIteratorSourceLocator.php [hunk 15]
```diff
@@ -28,11 +29,16 @@ class FileIteratorSourceLocator implements SourceLocator
...
+     */
+    private $astLocator;
+
+    /**
      * @param \Iterator|\SplFileInfo[] $fileInfoIterator note: only \SplFileInfo allowed in this iterator
      *
      * @throws InvalidFileInfo In case of iterator not contains only SplFileInfo
      */
-    public function __construct(Iterator $fileInfoIterator)
+    public function __construct(Iterator $fileInfoIterator, Locator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/FileIteratorSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/PhpInternalSourceLocator.php [hunk 16]
```diff
@@ -17,10 +20,11 @@
      */
     private $stubber;
 
-    public function __construct()
+    public function __construct(Locator $astLocator, Parser $parser)
```
Comment (Ocramius) [inline src/SourceLocator/Type/PhpInternalSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/SingleFileSourceLocator.php [hunk 17]
```diff
@@ -20,19 +21,18 @@ class SingleFileSourceLocator extends AbstractSourceLocator
...
      */
-    private $filename;
+    private $fileName;
 
     /**
-     * @param string $filename
      * @throws \Roave\BetterReflection\SourceLocator\Exception\InvalidFileLocation
      */
-    public function __construct(string $filename)
+    public function __construct(string $fileName, Locator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/SingleFileSourceLocator.php]: BC Break to be documented

File: src/SourceLocator/Type/StringSourceLocator.php [hunk 18]
```diff
@@ -22,22 +23,22 @@ class StringSourceLocator extends AbstractSourceLocator
     private $source;
 
     /**
-     * @param string $source
      * @throws \Roave\BetterReflection\SourceLocator\Exception\EmptyPhpSourceCode
      */
-    public function __construct(string $source)
+    public function __construct(string $source, Locator $astLocator)
```
Comment (Ocramius) [inline src/SourceLocator/Type/StringSourceLocator.php]: BC Break to be documented

File: src/Util/FindReflectionOnLine.php [hunk 19]
```diff
@@ -24,21 +23,25 @@
      */
     private $sourceLocator;
 
-    public function __construct(SourceLocator $sourceLocator)
+    /**
+     * @var Locator
+     */
+    private $astLocator;
+
+    public function __construct(SourceLocator $sourceLocator, Locator $astLocator)
```
Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: BC Break to be documented

File: src/Util/FindReflectionOnLine.php [hunk 20]
```diff
@@ -24,21 +23,25 @@
...
     {
         $this->sourceLocator = $sourceLocator;
+        $this->astLocator    = $astLocator;
     }
 
     /**
-     * @return self
+     * @deprecated
```
Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: Should we keep this method now that we have a global entry point for the library?

File: src/Reflector/ClassReflector.php [hunk 21]
```diff
@@ -28,15 +25,13 @@ public function __construct(SourceLocator $sourceLocator)
     }
 
     /**
-     * @return self
+     * @deprecated
```
Comment (Ocramius) [inline src/Reflector/ClassReflector.php]: Should we keep this method now that we have a global entry point for the library?

**Comment by Ocramius** [comment]
Note: test suite fails due to 1Gb memory usage - should I just raise that?

**Comment by Ocramius** [comment]
Hmm, actually... it shouldn't fail, as I'm really not doing anything heavyweight with it.

**Comment by Ocramius** [comment]
@kukulich can you check this patch if you got time?

File: src/Reflection/ReflectionFunctionAbstract.php [hunk 22]
```diff
@@ -52,8 +54,14 @@
...
     private $reflector;
 
+    /**
+     * @var Parser
+     */
+    private $parser;
+
     protected function __construct()
     {
+        $this->parser = (new Configuration())->phpParser();
```
Comment (Ocramius) [inline src/Reflection/ReflectionFunctionAbstract.php]: This is probably the offending memory leak: each function has a new parser.

I can probably revert the changes in here, as it's not really needed in the constructor, and can probably even be static (only used in `setBodyFromClosure` and `setBodyFromString`)

**Comment by kukulich** [comment]
@Ocramius I will probably have time later today. Now I have to babysit :)

File: src/SourceLocator/Type/ClosureSourceLocator.php [hunk 23]
```diff
@@ -131,10 +131,12 @@ public function leaveNode(Node $node) : void
 
             /**
              * @return Node[]|null[]|null
+             *
+             * @throws TwoClosuresOnSameLine
              */
             public function getClosureNodes() : ?array
             {
-                /** @var Node\Expr\Closure[] $closureNodesDataOnSameLine */
+                /** @var Node\Expr\Closure[][] $closureNodesDataOnSameLine */
```
Comment (kukulich) [inline src/SourceLocator/Type/ClosureSourceLocator.php]: Should be `Node[][]`, see line 120.

File: src/Reflection/ReflectionFunctionAbstract.php (hunk 22 above)
Comment (kukulich) [reply src/Reflection/ReflectionFunctionAbstract.php]: Both methods are removed in https://github.com/Roave/BetterReflection/pull/339

File: src/Configuration.php (hunk 3 above)
Comment (kukulich) [reply src/Configuration.php]: Yes, that would be nice :)

File: src/Reflector/ClassReflector.php (hunk 21 above)
Comment (kukulich) [reply src/Reflector/ClassReflector.php]: I vote for removing.

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 9 above)
Comment (Ocramius) [reply src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Not a BC break: not in a released version.

File: src/Reflector/FunctionReflector.php (hunk 5 above)
Comment (Ocramius) [reply src/Reflector/FunctionReflector.php]: Documented.

File: src/SourceLocator/Ast/Locator.php (hunk 6 above)
Comment (Ocramius) [reply src/SourceLocator/Ast/Locator.php]: Documented.

File: src/SourceLocator/Reflection/SourceStubber.php (hunk 7 above)
Comment (Ocramius) [reply src/SourceLocator/Reflection/SourceStubber.php]: Documented.

File: src/SourceLocator/Type/AbstractSourceLocator.php (hunk 8 above)
Comment (Ocramius) [reply src/SourceLocator/Type/AbstractSourceLocator.php]: Documented.

File: src/SourceLocator/Type/ClosureSourceLocator.php (hunk 11 above)
Comment (Ocramius) [reply src/SourceLocator/Type/ClosureSourceLocator.php]: Documented.

File: src/SourceLocator/Type/ComposerSourceLocator.php (hunk 12 above)
Comment (Ocramius) [reply src/SourceLocator/Type/ComposerSourceLocator.php]: Documented.

File: src/SourceLocator/Type/DirectoriesSourceLocator.php (hunk 13 above)
Comment (Ocramius) [reply src/SourceLocator/Type/DirectoriesSourceLocator.php]: Documented.

File: src/SourceLocator/Type/EvaledCodeSourceLocator.php (hunk 14 above)
Comment (Ocramius) [reply src/SourceLocator/Type/EvaledCodeSourceLocator.php]: Documented.

File: src/SourceLocator/Type/FileIteratorSourceLocator.php (hunk 15 above)
Comment (Ocramius) [reply src/SourceLocator/Type/FileIteratorSourceLocator.php]: Documented.

File: src/SourceLocator/Type/PhpInternalSourceLocator.php (hunk 16 above)
Comment (Ocramius) [reply src/SourceLocator/Type/PhpInternalSourceLocator.php]: Documented.

File: src/SourceLocator/Type/SingleFileSourceLocator.php (hunk 17 above)
Comment (Ocramius) [reply src/SourceLocator/Type/SingleFileSourceLocator.php]: Documented.

File: src/SourceLocator/Type/StringSourceLocator.php (hunk 18 above)
Comment (Ocramius) [reply src/SourceLocator/Type/StringSourceLocator.php]: Documented.

**Comment by Ocramius** [comment]
@asgrim this is ready to be :ship: -ed

File: appveyor.yml [hunk 24]
```diff
@@ -58,6 +58,6 @@ install:
   - IF %dependencies%==highest composer update --prefer-dist --no-interaction --no-progress
 test_script:
   - ps: cd $Env:project_directory
-  - vendor\bin\phpunit
+  - php -dmemory_limit=2G vendor\bin\phpunit
```
Comment (kukulich) [inline appveyor.yml]: `vendor\bin\phpunit` is `.bat` file on Windows.

Use `php -dmemory_limit=2G vendor\phpunit\phpunit\phpunit`

File: README.md [hunk 25]
```diff
@@ -54,6 +54,23 @@ $classInfo = ReflectionClass::createFromName('Foo\Bar\MyClass');
...
+ * `Roave\BetterReflection\SourceLcoator\Type\AutoloadSourceLocator::__construct` now has mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\ClosureSourceLocator::__construct` has additional mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\ComposerSourceLocator::__construct` has additional mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\DirectoriesSourceLocator::__construct` has additional mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\EvaledCodeSourceLocator::__construct` now has mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\FileIteratorSourceLocator::__construct` has additional mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\PhpInternalSourceLocator::__construct` now has mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\SingleFileSourceLocator::__construct` has additional mandatory constructor arguments
+ * `Roave\BetterReflection\SourceLcoator\Type\StringSourceLocator::__construct` has additional mandatory constructor arguments
```
Comment (kukulich) [inline README.md]: Typo

File: README.md [hunk 26]
```diff
@@ -54,6 +54,23 @@ $classInfo = ReflectionClass::createFromName('Foo\Bar\MyClass');
...
+ * `Roave\BetterReflection\Util\FindReflectionOnLine::__construct` now has mandatory constructor arguments
+ * `Roave\BetterReflection\Reflector\ClassReflector::buildDefaultReflector()` was dropped, please use the newly
+   introduced `Roave\BetterReflection\BetterReflection#classReflector()` instead
```
Comment (kukulich) [inline README.md]: Would not be better to use `::` instead of `#` ?

Comment (Ocramius) [reply README.md]: `#` for instance methods, `::` for static methods

File: test/unit/Reflection/ReflectionFunctionAbstractTest.php [hunk 27]
```diff
@@ -36,6 +40,32 @@
...
+    /**
+     * @var Locator
+     */
+    private $astLocator;
+
+    protected function setUp() : void
+    {
+        parent::setUp();
+
+        $configuration = (new BetterReflection());
```
Comment (kukulich) [inline test/unit/Reflection/ReflectionFunctionAbstractTest.php]: Useless parentheses.

**Comment by Ocramius** [comment]
:ship: :running_man:

**Comment by TomasVotruba** [comment]
Is there any place I could found before/after code snippets? I'd like to migrete to this

**Comment by Ocramius** [comment]
I documented the new approach in the "how to use" docs

**Comment by TomasVotruba** [comment]
The diff is much better for me, since I already use the tool.

**Comment by Ocramius** [comment]
Open an issue please

**Comment by TomasVotruba** [comment]
Oh no, I have no issue anymore. I'm very happy with the diff :)

**Comment by TomasVotruba** [comment]
Just wondering, how many hours did it take you all this? There is lot of work

**Comment by Ocramius** [comment]
I didn't count: I'm at a conference. I'd say ~4 hours if you don't count
interruptions

**Pull request merged**